Replace Create React App (CRA) with Vite (CRASM-2750)

hawkishpolicy opened the pull request:
<!-- GitHub renders PRs such that soft line breaks are treated as hard
line breaks.  In order to make this template render as expected we
therefore have to avoid soft breaks and therefore will offend
markdownlint with long lines.  This is the reason for the markdownline
disable directive just below.

For more details see:
https://github.com/github/markup/issues/1050#issuecomment-294654762 -->
<!-- markdownlint-disable MD013 -->
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##
- This commit includes the following changes:
  - Migration from Create React App (CRA) to Vite for the frontend.
  - Refactoring of frontend build and development scripts.
  - Addition of new TypeScript definitions for images, SCSS, and SVG files.
  - Creation of import alias for frontend source files.
  - Relocation of `index.html` to the root of the `frontend` directory.
  - Renaming of environment files to follow Vite conventions.
    - REACT_APP_ prefixed variables are now prefixed with VITE_.
    - Ensure those variables are called correctly throughout the codebase.
  - Replacement of Jest with Vitest for testing.
    - Updated test files to use Vitest syntax and features.
  - Updated ESLint configuration to work with Vite and TypeScript.
  - Adjusted Dockerfile and docker-compose configurations for Vite.
  - Added src/frontend to "include:" section of tsconfig.json. at root.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##
- Closes CRASM-2750
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [ ] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [ ] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [ ] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.
- [ ] Bump major, minor, patch, pre-release, and/or build versions [as appropriate](https://semver.org/#semantic-versioning-specification-semver) via the `bump_version` script *if* this repository is versioned *and* the changes in this PR [warrant a version bump](https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api).
- [ ] Create a pre-release (necessary if and only if the pre-release version was bumped).

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release (necessary if and only if the version was bumped).

hawkishpolicy committed:
Large initial commit to replace CRA with Vite

- This commit includes the following changes:
  - Migration from Create React App (CRA) to Vite for the frontend.
  - Refactoring of frontend build and development scripts.
  - Addition of new TypeScript definitions for images, SCSS, and SVG files.
  - Creation of import alias for frontend source files.
  - Relocation of `index.html` to the root of the `frontend` directory.
  - Renaming of environment files to follow Vite conventions.
    - REACT_APP_ prefixed variables are now prefixed with VITE_.
    - Ensure those variables are called correctly throughout the codebase.
  - Replacement of Jest with Vitest for testing.
    - Updated test files to use Vitest syntax and features.
  - Updated ESLint configuration to work with Vite and TypeScript.
  - Adjusted Dockerfile and docker-compose configurations for Vite.
  - Added src/frontend to "include:" section of tsconfig.json. at root.

hawkishpolicy committed:
Update GitHub Actions workflows to Node.js 20.19.0

hawkishpolicy committed:
Update prettier formatting

- in frontend documentation

hawkishpolicy committed:
Remove react-app preset from ESLint configuration

- Updated the ESLint configuration to remove the `react-app` preset, which is no longer needed after migrating from Create React App (CRA) to Vite. This change ensures that the ESLint setup is compatible with the new Vite environment.

hawkishpolicy committed:
Update eslint plugin to latest version

hawkishpolicy committed:
Install 'eslint-plugin-react-hooks' for linting

hawkishpolicy committed:
Remove CRA related dependencies and overides:

- Webpack and related plugins
- Babel loader and related plugins
- Jest testing library and related configurations
- removed react-app and react-app/jest from eslint config

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Replace-CRA-with-Vite-CRASM-2750

hawkishpolicy committed:
Remove testing-library/jest-dom package

hawkishpolicy committed:
Updated imports for testing

- restored '@testing-library/jest-dom' import in setupTests.ts
- brought in 'vitest' imports in various test files

hawkishpolicy committed:
updated imports in test files

hawkishpolicy committed:
Updated vitest imports in routeGuard.spec.tsx

hawkishpolicy committed:
Clear error in searchBar.spec.tsx

- added missing initialValue prop to SearchBar component in test.

hawkishpolicy committed:
Reorganize vite-type files & upgrade node to 22.17

hawkishpolicy committed:
Remove CRA eslint dep from root package.json

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Replace-CRA-with-Vite-CRASM-2750

hawkishpolicy committed:
Update frontend dependencies & type definitions

- Ran npm install after merging the latest changes from the develop branch.
- Added type definitions to "defaultValues" in RegisterForm.tsx to resolve TypeScript errors.

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Replace-CRA-with-Vite-CRASM-2750

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Replace-CRA-with-Vite-CRASM-2750

hawkishpolicy committed:
Update node to 22.17 in GitHub workflows

aloftus23 commented on frontend/package.json:
```diff
@@ -111,32 +110,24 @@
+    "vitest": "^3.2.4"
   },
   "engines": {
     "node": ">=18.0.0"
```
Need to update the engines --> node to:
"node": ">=22.17.0"

**This should also be done in docs/package.json**

hawkishpolicy replied:
Done.

aloftus23 commented on .github/workflows/security.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
       - uses: actions/setup-node@v3
         with:
-          node-version: '18'
+          node-version: '20.19'
```
Might as well change all of node versions in this file to 22.17

hawkishpolicy replied:
Done.

aloftus23 commented on .github/workflows/docs.yml:
```diff
@@ -28,7 +28,7 @@ jobs:
       - uses: actions/setup-node@v3
         with:
-          node-version: '18'
+          node-version: '20.19'
```
```suggestion
          node-version: '22.17'
```

hawkishpolicy replied:
Done.

aloftus23 reviewed (commented):
See below places that still need update to node 22.

Last place is the frontend/serverless.yml. Line 64:   runtime: nodejs16.x
Should be updated to   runtime: nodejs22.x

I can explain on a call but this is the definition of the lambda that the frontend app is spun up in AWS.

hawkishpolicy committed:
Update Node.js version in workflows & package.json

- Update Node.js version in GitHub workflows/docs, workflows/security and frontend & docs package.json files to 22.17.0.

hawkishpolicy committed:
NPM audit fix to address vuln with form-data:

- form-data uses unsafe random function in form-data for choosing boundary - https://github.com/advisories/GHSA-fjxv-7rqg-78g4

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Replace-CRA-with-Vite-CRASM-2750

aloftus23 commented:
> See below places that still need update to node 22.
> 
> Last place is the frontend/serverless.yml. Line 64: runtime: nodejs16.x Should be updated to runtime: nodejs22.x
> 
> I can explain on a call but this is the definition of the lambda that the frontend app is spun up in AWS.

@hawkishpolicy Just one more sport you still need to go to Node22:
frontend/serverless.yml. Line 64: runtime: nodejs16.x
Should be updated to runtime: nodejs22.x

jayjaybunce commented on dev.env.example:
```diff
@@ -16,8 +16,12 @@ MDL_SECONDARY_NAME=crossfeed_mini_datalake22
-REACT_APP_FARGATE_LOG_GROUP=crossfeed-staging-worker
+MDL_SECONDARY_NAME=crossfeed_mini_datalake22
+
+LOCAL_MDL_USERNAME=
```
Remove this duplicate. `LOCAL_MDL_USERNAME` already defined on line 17.

hawkishpolicy replied:
Removed duplicate variables MDL_SECONDARY_NAME and LOCAL_MDL_USERNAME

jayjaybunce reviewed (commented):
LGTM, left one comment for a duplicate env variable.

Testing functionality locally, we will need to update some params in parameter store or add new ones to support both REACT_APP params and VITE params until the changes in the PR make it across all environments.

hawkishpolicy committed:
Update serverless.yml to use Node.js 22.x runtime

hawkishpolicy committed:
Remove duplicate variables in dev.env.example

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Replace-CRA-with-Vite-CRASM-2750

hawkishpolicy committed:
Fixed breaking changes from merge

- Updated AuthLogin and OktaCallback to use Vite environment variables.

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Replace-CRA-with-Vite-CRASM-2750

schmelz21 reviewed (approved):
Test Team please take in consideration the following:
Environment: All deployment and local .env files must be updated to use VITE_ prefixes.
Testing: Ensure all Vitest tests run as expected and coverage is sufficient.
Asset Loading: Asset paths were updated to work with Vite (may need checking on static files, images, CSS).
Validate that backend and frontend communicate as expected.
Check that all documentation builds/render correctly and is accurate.

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Replace-CRA-with-Vite-CRASM-2750

hawkishpolicy commented:
> > See below places that still need update to node 22.
> > Last place is the frontend/serverless.yml. Line 64: runtime: nodejs16.x Should be updated to runtime: nodejs22.x
> > I can explain on a call but this is the definition of the lambda that the frontend app is spun up in AWS.
> 
> @hawkishpolicy Just one more sport you still need to go to Node22: frontend/serverless.yml. Line 64: runtime: nodejs16.x Should be updated to runtime: nodejs22.x

Resolved issue with https://github.com/cisagov/XFD/pull/1017/commits/e8418242c0941780ef75e0a1e4fc45ed2fc93e3d
